### DISCUSSION TITLE
Mocking a federated apollo-server 3

### DISCUSSION THREAD

**Discussion by setchy** [apollographql/apollo-server#5507] [Q&A] [answered]
I've been able to migrate our federated gateway from apollo-server 2 to 3 this weekend without too much drama.  

All appears to be working with the exception of being able to use mocks with gateway (which was previously functioning using apollo-server-2)

I've read the [mocking section](https://www.apollographql.com/docs/apollo-server/migration/#mocking) of the migration guide and the [apollo-server v3 mocking page](https://www.apollographql.com/docs/apollo-server/testing/mocking/) dozens of times and yet still can't work out what I'm missing.

Any pointers would be greatly appreciated.

I'm working with the apollo-server-express variety

**Comment by glasser** [comment, marked as answer]
What do you mean by "use mocks with gateway"? Are you trying to mock out your entire federated service? Or to mock out individual subgraphs behind the gateway? What's not working?

**Reply by setchy** [reply]
We have a separate deployment of our federated gateway (managed federation) which:
* mocks the entire federated schema using a combination of `faker.js` and more intelligent business-centric mocks
* disables JWT auth to make it easier to access
* GraphQL Playground (will replace with Apollo Sandbox during AS3 migration) 
* GraphQL Voyager - visual representation of our connected types

To help explain and/or triage I put together [setchy/apollo-federation-mocked-examples](https://github.com/setchy/apollo-federation-mocked-examples) repo earlier today which contains 4 different examples of mocking `apollo-server`
1. Apollo Server 2 with managed federation
2. Apollo Server 2 without managed federation
3. Apollo Server 3 with managed federation
4. Apollo Server 3 without managed federation

Pattern 1 is what we're using (albeit the express variant) and appears to be the only combination that actually mocks the entire gateway schema.  In AS3 it always returns `null` and without managed federation the subgraph services appear to always be called regardless of mock configuration

**Reply by glasser** [reply]
Hmm, I wasn't able to get 1 working either because I think there's a missing step about publishing subgraph schemas to studio?

Yeah, it does look like we changed the implementation of `mocks` in AS3 to only apply in the no-gateway case. I'm actually surprised that it had any effect at all in AS2, since normally the resolver functions on the schema are entirely ignored when you're using the gateway as your executor! I didn't get 1 working during the time I had to work on this today, but I'd be curious to understand why it did work...

What I'm curious about is what you're trying to do here. Is the idea that you want to use managed federation to *fetch the schema*, but you then want to execute it using mocks instead of using actual federation execution?  In that case, I think we might just want to entirely disconnect the gateway object from the ApolloServer, or make your own GatewayInterface implementation.  Eg:

```js

const gateway = new ApolloGateway({});
const { schema } = await gateway.load();

const server = new ApolloServer({
  schema,
  mocks,
});
```

This will not change the schema over time as the schema changes, though. To do that you can implement your own gateway subclass that applies mock functions to received schema and uses executor to run graphql execute I guess. We could perhaps ship such an implementation of GatewayInterface in `@apollo/gateway`...

**Reply by setchy** [reply]
You're correct - I had assumed that for the provided `APOLLO_KEY` and `APOLLO_GRAPH_REF` pair that at least one subgraph had been previously published to the Studio schema registry.

Your summary is pretty spot on.  

To add further color to it, our use-case is limited to a special `sandbox` environment whereby we:
* used during certain internal training / education awareness sessions
* product prototyping against an always up-to-date supergraph (managed federation) with quick mocked responses and no downstream service dependencies except for the gateway
* mock entire schema surface area with a combination of 
** purely random data (faker.js) where we don't care much about what it is, or
** domain-appropriate static mocks for certain types / fields
* relax all authn, authz, entitlements layers 

I know there is support in Apollo Studio / Apollo Sandbox to feature flag mocked responses off/on, but that is only helpful when staying inside the Apollo Studio ecosystem.  Or [Workbench](https://www.apollographql.com/blog/backend/federation/apollo-workbench-a-better-way-to-design-federated-graphs/#mocking-your-design) to run a local gateway with mocks (whats documented in that blog post is very similar to what we do re: custom mocks, just ours is at a single managed gateway level)

**Reply by glasser** [reply]
I threw this together for forthcoming AS 3.1.0. How does that look to you? See the sample code in the PR description. https://github.com/apollographql/apollo-server/pull/5539

**Reply by setchy** [reply]
Awesome, thanks @glasser - looks like this will work for our needs
